examples/grad3d/grad3d.f90

```diff
@@ -0,0 +1,625 @@
+         df(i,j,nz) = coeff*(ff(i,j,1)-ff(i,j,nz-1))
+      enddo
+      !$acc end kernels
+
+   end subroutine derz 
+   !=====================================================================
+   ! Calculate divergence using halo-cell exchange (data in X-pencil)
+   !=====================================================================
+   subroutine test_derX(df)
+
+      implicit none
+      ! Arguments
+      real(mytype), intent(out), dimension(xsize(1),xsize(2),xsize(3)) :: df
+
+      integer :: i, j, k
+      real(mytype), parameter :: twopi=2._mytype*acos(-1._mytype)
+      real(mytype) :: x, y, z
+      real(mytype) :: dphi, dphi_num
+      real(mytype) :: error = 0._mytype
+      real(mytype) :: err_all = 0._mytype
+      real(mytype) :: dphi2 = 0._mytype
+      real(mytype) :: sum_dphi2 = 0._mytype
+      integer :: xe1, xe2, xe3
+      integer :: xs1, xs2, xs3
+
+      xe1 = xsize(1)
+      xe2 = xsize(2)
+      xe3 = xsize(3)
+      xs1 = xstart(1)
+      xs2 = xstart(2)
+      xs3 = xstart(3)
+
+      ! Initial velocity
+      !$acc parallel loop default(present) reduction(+:error)
+      do k=1,xe3
+        do j=1,xe2
+          do i=1,xe1
+             z = (k + xs3 - 2) * dz
+             y = (j + xs2 - 2) * dy
+             x = (i + xs1 - 2) * dx
+             dphi = -2._mytype * twopi * sin(twopi * (x / lx)) * cos(twopi * (y / ly)) * sin(twopi * (z / lz))
+             dphi2 = dphi2 + dphi * dphi
+             dphi_num = df(i,j,k)
+             error = error + (dphi-dphi_num)*(dphi-dphi_num)
+          enddo
+        enddo
+      enddo
+      !$acc end loop
+      call MPI_ALLREDUCE(error, err_all, 1, real_type, MPI_SUM, MPI_COMM_WORLD, ierror)
+      call MPI_ALLREDUCE(dphi2, sum_dphi2, 1, real_type, MPI_SUM, MPI_COMM_WORLD, ierror)
+      err_all = sqrt(err_all/sum_dphi2) / (real(nx, mytype) * real(ny, mytype) * real(nz, mytype))
+      error_ref = err_all
+      
+      if (nrank == 0) then
+         write (*, *) 'DX error / mesh point: ', err_all
+      endif
+
+   end subroutine test_derX
+
+   !=====================================================================
+   ! Calculate divergence using halo-cell exchange (data in Y-pencil)
+   !=====================================================================
+   subroutine test_derY(df)
+
+      implicit none
+      ! Arguments
+      real(mytype), intent(out), dimension(xsize(1),xsize(2),xsize(3)) :: df
+
+      integer :: i, j, k
+      real(mytype), parameter :: twopi=2._mytype*acos(-1._mytype)
+      real(mytype) :: x, y, z
+      real(mytype) :: dphi, dphi_num
+      real(mytype) :: error = 0._mytype
+      real(mytype) :: err_all = 0._mytype
+      real(mytype) :: dphi2 = 0._mytype
+      real(mytype) :: sum_dphi2 = 0._mytype
+      integer :: xe1, xe2, xe3
+      integer :: xs1, xs2, xs3
+
+      xe1 = xsize(1)
+      xe2 = xsize(2)
+      xe3 = xsize(3)
+      xs1 = xstart(1)
+      xs2 = xstart(2)
+      xs3 = xstart(3)
+
+      ! Initial velocity
```

Comment:
Compute the error against the analytical solution

Reply:
done